app/services/SubscriptionService.scala

```diff
@@ -18,42 +18,169 @@ package services
-    hc: HeaderCarrier,
-    ec: ExecutionContext
-  ): Future[Either[ApiError, SubscriptionID]] =
-    getDisplaySubscriptionId(safeID) flatMap {
-      case Some(subscriptionID) =>
-        EitherT.rightT(subscriptionID).value
+  def getContactDetails(
+    subscriptionId: SubscriptionID,
+    userAnswers: UserAnswers
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[UserAnswers]] =
+    getSubscription(subscriptionId) map {
+      responseOpt =>
+        responseOpt.flatMap {
+          responseDetail => populateUserAnswers(responseDetail.success, userAnswers)
+        }
+    }
+
+  def updateContactDetails(
+    subscriptionId: SubscriptionID,
+    userAnswers: UserAnswers
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] =
+    getSubscription(subscriptionId).flatMap {
+      case Some(displaySubscriptionResponse) =>
+        UpdateSubscriptionRequest.convertToRequest(displaySubscriptionResponse, userAnswers) match {
+          case Some(updateSubscriptionRequest) => subscriptionConnector.updateSubscription(updateSubscriptionRequest)
+          case _ =>
+            logger.warn("updateContactDetails: failed to convert userAnswers to RequestDetailForUpdate")
+            Future.successful(false)
+        }
       case _ =>
-        (CreateSubscriptionRequest.buildSubscriptionRequest(safeID, userAnswers, affinityGroup) match {
-          case Some(subscriptionRequest) =>
-            val response = subscriptionConnector.createSubscription(subscriptionRequest)
+        logger.warn("updateContactDetails: readSubscription call failed to fetch data")
+        Future.successful(false)
+    }
 
-            response
+  def contactDetailsChanged(
+    subscriptionId: SubscriptionID,
+    userAnswers: UserAnswers
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[Boolean]] =
+    getSubscription(subscriptionId).map {
+      case Some(subscriptionDisplayResponse) =>
+        val subscription = subscriptionDisplayResponse.success
+        val secondaryContact =
+          (userAnswers.get(OrganisationHaveSecondContactPage), subscription.secondaryContact, userAnswers.get(OrganisationSecondContactNamePage)) match {
+            case (Some(true), _, Some(name)) =>
+              populateResponseDetails[ChangeOrganisationSecondaryContactDetailsPages](
+                userAnswers,
+                OrganisationDetails(name)
+              )
+            case (Some(true), Some(contactInformation), _) =>
+              populateResponseDetails[ChangeOrganisationSecondaryContactDetailsPages](
+                userAnswers,
+                contactInformation.contactInformation
+              )
+            case _ => None
+          }
+
+        for {
+          primaryContact <- populateResponseDetails[ChangeOrganisationPrimaryContactDetailsPages](
+            userAnswers,
+            subscription.primaryContact.contactInformation
+          )
+        } yield !subscriptionDisplayResponse
```

Comment:
I'm struggling to understand what this is doing which is a concern. Can we go through it and see if we can either simplify it or perhaps break it down and use variables to help signpost whats going on?

Reply:
I've simplified the method using `vals` and methods and hopefully should be clearer now. The method does the below:

1. gets the current subscription detail from the backend
2. gets the primary contact details which the user supplied in the user answers
3. gets the secondary contact details which the user supplied in the user answers
4. compares the primary contact from the user answers with the one from the backend for changes
5. compares the secondary contact from the user answers with the one from the backend for changes
6. the result of the previous comparisons is used to display the `Check` version of the page